src/tools/rbd_mirror/PoolReplayer.cc

```diff
@@ -677,6 +795,20 @@ void PoolReplayer<I>::update_namespace_replayers() {
           ctx->complete(r);
         });
     namespace_replayer->init(on_init);
+    {
+      C_SaferCond add_peer_cond;
+      auto add_peer_gather_ctx = new C_Gather(cct, &add_peer_cond);
+      for (const auto& kv : m_remote_peers_ressource) {
+        // add peers
+        namespace_replayer->add_peer(kv.second.peer,
+                                     add_peer_gather_ctx->new_sub());
```

Comment:
so we call add_peer when init is still in process. Is it safe?

Reply:
Yes it should be safe, peers can be added to the InstanceReplayer at the init if the list of peers is not empty or later via the add_peer chain (the add_peer of the Namespace replayer also check if the InstanceReplayer do exists).

But because of this logic I sometimes have two add_peer on the InstanceReplayer for the same peer that's why I am checking there if the push operation to the set succeed instead of the usual assert. Maybe I will refactor a bit this later on when doing the unit test.